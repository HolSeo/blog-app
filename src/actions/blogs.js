import database from '../firebase/firebase'

export const addBlog= (blog) => ({
    type: 'ADD_BLOG',
    blog
})

export const startAddBlog = (blogData = {}) => {
    return (dispatch,getState) => {
        const uid = getState().auth.uid
        const {
            title =  '',
            body = '',
            createdAt = 0
        } = blogData
        const blog = { title, body, createdAt }
        return database.ref(`users/${uid}/blogs`).push(blog).then((ref) => {
            dispatch(addBlog({
                id: ref.key,
                ...blog
            }))
        })
    }
}

export const editBlog = (id,updates) => ({
    type: 'EDIT_BLOG',
    id,
    updates
})

export const startEditBlog = (id, updates) => {
    return (dispatch,getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/blogs/${id}`).update(updates).then(() => {
            dispatch(editBlog(id,updates))
        })
    }
}

export const setBlogs = (blogs) => ({
    type: 'SET_BLOGS',
    blogs
})

export const startSetBlogs = () => {
    return (dispatch,getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/blogs`).once('value').then((snapshot) => {
            const blogs = []
            snapshot.forEach((childSnapshot) => {
                blogs.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            })
            dispatch(setBlogs(blogs))
        })
    }
}