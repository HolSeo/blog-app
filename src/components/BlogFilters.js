import React from 'react'

export const BlogFilters = () => (
    <div>
        <input type="text" placeholder="Search posts"/>
        <select>
            <option>By Title</option>
            <option>By Date</option>
        </select>
    </div>
)

export default BlogFilters