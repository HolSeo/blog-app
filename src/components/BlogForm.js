import React from 'react'
import 'react-dates/initialize'
import { SingleDatePicker } from 'react-dates'
import moment from 'moment'

export class BlogForm extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            title: this.props.blog ? this.props.blog.title : '',
            body: this.props.blog ? this.props.blog.body : '',
            createdAt: this.props.blog ? moment(this.props.blog.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        }
    }
    onTitleChange = (e) => {
        const title = e.target.value
        this.setState(() => ({ title }))
    }
    onBodyChange = (e) => {
        const body = e.target.value
        this.setState(() => ({ body }))
    }
    onDateChange = (createdAt) => {
        this.setState(() => ({ createdAt }))
    }
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    }
    onSubmit = (e) => {
        e.preventDefault()
        if (!this.state.title || !this.state.body) {
            this.setState(() => ({ error: 'Please fill out Title AND Body' }))
        } else {
            this.setState(() => ({ error: '' }))
            this.props.onSubmit({
                title: this.state.title,
                body: this.state.body,
                createdAt: this.state.createdAt.valueOf()
            })
        }
    }
    render() {
        return (
            <form onSubmit={this.onSubmit}>
                {this.state.error && <p>{this.state.error}</p>}
                <input 
                    type="text"
                    placeholder="Title"
                    onChange={this.onTitleChange}
                    value={this.state.title}
                />
                <SingleDatePicker 
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                />
                <textarea
                    placeholder="Body"
                    onChange={this.onBodyChange}
                    value={this.state.body}
                >
                </textarea>
                <button>Save Blog</button>
            </form>
        )
    }
}

export default BlogForm










































// import React from 'react'
// import moment from 'moment'
// import 'react-dates/initialize'
// import { SingleDatePicker } from 'react-dates'

// export class BlogForm extends React.Component {
//     state = {
//         title: '',
//         body: '',
//         createdAt: moment(),
//         calendarFocused: false,
//         error: ''
//     }

//     onTitleChange = (e) => {
//         const title = e.target.value
//         this.setState(() => ({ title }))
//     }
//     onBodyChange = (e) => {
//         const body = e.target.value
//         this.setState(() => ({ body }))
//     }
//     onDateChange = (createdAt) => {
//         this.setState(() => ({ createdAt }))
//     }
//     onFocusChange = ({ focused }) => {
//         this.setState(() => ({ calendarFocused: focused}))
//     }
//     onSubmit = (e) => {
//         e.preventDefault()
//         if (!this.state.description || !this.state.amount) {
//             this.setState(() => ({ error: 'Please provide description and amount' }))
//         } else {
//             this.setState(() => ({ error: '' }))
//             this.props.onSubmit({
//                 title: this.state.title,
//                 body: this.state.body,
//                 createdAt: this.state.createdAt.valueOf()
//             })
//         }
//     }

//     render() {
//         return (
//             <form onSubmit={this.onSubmit}>
//                 {this.state.error && <p>{this.state.error}</p>}
//                 <input
//                     type="text"
//                     placeholder="Title"
//                     autoFocus
//                     value={this.state.title}
//                     onChange={this.onTitleChange}
//                 />
//                 <SingleDatePicker
//                     date={this.state.createdAt}
//                     onDateChange={this.onDateChange}
//                     focused={this.state.calendarFocused}
//                     onFocusChange={this.onFocusChange}
//                     numberOfMonths={1}
//                     isOutsideRange={() => false}
//                 />
//                 <textarea
//                     placeholder="Add the body of your blog."
//                     onChange={this.onBodyChange}
//                     value={this.state.body}
//                 >
//                 </textarea>
//                 <button>Save Blog</button>
//             </form>
//         )
//     }
// }

// export default BlogForm