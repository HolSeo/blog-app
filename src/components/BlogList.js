import React from 'react'
import { connect } from 'react-redux'
import BlogListItem from './BlogListItem'
import getVisibleBlogs from '../selectors/getVisibleBlogs'


export const BlogList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Blogs</div>
            <div className="show-for-desktop">Blogs</div>
            <div className="show-for-desktop">Author</div>
        </div>
        <div className="list-body">
            {
                props.blogs.length === 0 ? (
                    <div className="list-item list-item--message">
                        <span>No Blogs</span>
                    </div>
                ) : (
                    props.blogs.map((blog) => 
                    <BlogListItem key={blog.id} {...blog} />)
                )
            }
        </div>
    </div>
)

const mapStateToProps = (state) => {
    return {
        blogs: getVisibleBlogs(state.blogs,state.filters)
    }
}

export default connect(mapStateToProps)(BlogList)


