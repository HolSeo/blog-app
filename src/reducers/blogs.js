const defaultblogsReducerState = []

const blogsReducer = (state = defaultblogsReducerState, action) => {
    switch (action.type) {
        case 'ADD_BLOG':
            return [...state, action.blog]
        default:
            return state
    }
}

export default blogsReducer