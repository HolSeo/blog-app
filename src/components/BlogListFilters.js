import React from 'react'
import { connect } from 'react-redux'
import { DateRangePicker } from 'react-dates'
import { setStartDate, setEndDate, setTextFilter } from '../actions/filters'

export class BlogListFilters extends React.Component {
    state = {
        calendarFocused: null
    }
    onChange = (e) => {
        const text = e.target.value
        this.props.setTextFilter(text)
    }
    onDatesChange = ({ startDate,endDate }) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onFocusChange = (focusedInput) => {
        this.setState(() => ({ calendarFocused: focusedInput }))
    }
    render() {
        return (
            <div>
                <input 
                    type="text" 
                    placeholder="search" 
                    onChange={this.onChange}
                    value={this.props.filters.text}
                />
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    startDateId="startDate"
                    endDate={this.props.filters.endDate}
                    endDateId="endDate"
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters
})

const mapDispatchToProps = (dispatch) => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setTextFilter: (text) => dispatch(setTextFilter(text))
})

export default connect(mapStateToProps, mapDispatchToProps)(BlogListFilters)