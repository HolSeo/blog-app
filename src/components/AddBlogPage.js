import React from 'react'
import { connect } from 'react-redux'
import { startAddBlog } from '../actions/blogs'
import BlogForm from './BlogForm'

export class AddBlogPage extends React.Component {
    onSubmit = (blog) => {
        this.props.startAddBlog(blog)
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
                Add Blog Page
                <BlogForm onSubmit={this.onSubmit}/>
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddBlog: (blog) => dispatch(startAddBlog(blog))
})

export default connect(undefined,mapDispatchToProps)(AddBlogPage)