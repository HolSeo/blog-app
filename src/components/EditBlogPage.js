import React from 'react'
import { connect } from 'react-redux'
import BlogForm from './BlogForm'
import { startEditBlog, startRemoveBlog } from '../actions/blogs'

export class EditBlogPage extends React.Component {
    onRemove = () => {
        this.props.startRemoveBlog(this.props.blog.id)
        this.props.history.push('/')
    }
    onSubmit = (blog) => {
        this.props.startEditBlog(this.props.blog.id,blog)
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
                <BlogForm 
                    blog={this.props.blog} 
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onRemove}>Remove Blog</button>
            </div>
        )
    }
}

const mapStateToProps = (state,props) => {
    return {
        blog: state.blogs.find((blog) => blog.id === props.match.params.id)
    }
}

const mapDispatchToProps = (dispatch) => ({
    startEditBlog: (id, updates) => dispatch(startEditBlog(id,updates)),
    startRemoveBlog: (id) => dispatch(startRemoveBlog(id))
})

export default connect(mapStateToProps,mapDispatchToProps)(EditBlogPage)

