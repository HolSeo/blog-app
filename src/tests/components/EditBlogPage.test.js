import React from 'react'
import { shallow } from 'enzyme'
import { EditBlogPage } from '../../components/EditBlogPage';
import blogs from '../fixtures/blogs'

let wrapper, startRemoveBlog, startEditBlog, history

beforeEach(() => {
    startRemoveBlog = jest.fn()
    startEditBlog = jest.fn()
    history = { push: jest.fn() }
    wrapper = shallow(<EditBlogPage 
        startRemoveBlog={startRemoveBlog} 
        startEditBlog={startEditBlog} 
        history={history}
        blog={blogs[0]}
    />)
})

test('should render EditBlogPage correctly', () => {
    expect(wrapper).toMatchSnapshot()
})

test('should call startEditBlog with correct arguments', () => {
    wrapper.find('BlogForm').prop('onSubmit')(blogs[0])
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(startEditBlog).toHaveBeenLastCalledWith(blogs[0].id, blogs[0])
})

test('should call startRemoveBlog correcty', () => {
    wrapper.find('button').simulate('click')
    expect(startRemoveBlog).toHaveBeenLastCalledWith(blogs[0].id)
    expect(history.push).toHaveBeenLastCalledWith('/')
})