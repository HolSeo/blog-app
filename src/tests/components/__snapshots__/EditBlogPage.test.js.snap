// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditBlogPage correctly 1`] = `
<div>
  <BlogForm
    blog={
      Object {
        "body": "This is first one!",
        "createdAt": 0,
        "id": "1",
        "title": "Title One",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove Blog
  </button>
</div>
`;
